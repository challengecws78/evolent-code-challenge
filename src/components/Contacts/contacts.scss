.tableContainer {
    padding-top: 1em;
    .contactTable {
        width: 60%;
        margin: 0 auto;
        padding: 10px;
    }
    .no-results {
        text-align: center;
    }
}
  
.contact-input {
    margin-top: 50px;
    button {
        display: block;
        margin: 0 auto;
    }
}

table {
    border: 1px solid #333;
    border-collapse: collapse;

    th {
        padding: 5px;
        margin: 0;
        text-align: center;
        font-weight: normal;
        font-size: 15px;
    }

    td {
        text-align: center;
        padding: 5px;
    }

    thead {
        background-color: #bbb;
        th {
            border-bottom: 1px solid #ccc;
        }
    } 
}

.contact-form {
    display: block;
    width: 300px;
    margin: 0 auto;
}

h1, h2 {
    text-align: center;
}

a {
    display: inline-block;
    padding: 20px;
    margin: 0 auto;
}

header {
    width: 100%;
    color: #fff;
    background: #333;
    nav {
        color: #fff;
        text-align: center;
    }
}

body {
    margin: 0px;
}

table tbody tr:hover {
    background-color: #fff;
}
  
table tbody tr:hover td:hover {
    background-color: #333;
    color: #fff;
}

footer {
    height: 40px;
    width: 100%;
    text-align: center;
    font-family: 'Rokkitt', serif;
    text-shadow: 2px 2px #FFF;
    font-size: 15px;
    margin-top: 10em;
    border-top: 1px solid #000;
    padding: 10px 0px 10px 0px;
}

.filters {
    margin-top: 2em; 
    fieldset{
        border: 1px double #000;
        width: 50%;
        margin: 0 auto;
        padding: 20px 0 20px 0;
        input, label {
            text-align: center;
            display: block;
            margin: 0 auto;
        }
        input {
            width: 50%;
        }
        button {
            display: block;
            margin: 0 auto;
        }
        .sort {
            display:flex;
            align-items: center;
            justify-content: center;
        }
    }
}

