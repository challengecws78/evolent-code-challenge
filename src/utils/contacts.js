export const contacts = [
    {
      "id": "1",
      "firstName": "Vergie",
      "lastName": "Vankirk",
      "email": "Vergie.Vankirk@gmailio.com",
      "phone": "(374) 684-6831",
      "status": "active"
    },
    {
      "id": "2",
      "firstName": "Milford",
      "lastName": "Mizzell",
      "email": "milfordmizzell@gmailio.com",
      "phone": "(505) 2812342",
      "status": "active"
    },
    {
      "id": "3",
      "firstName": "Magen",
      "lastName": "Mart",
      "email": "magen.mart@eggmail.com",
      "phone": "(757) 702-8058",
      "status": "active"
    },
    {
      "id": "4",
      "firstName": "Kieth",
      "lastName": "Kalper",
      "email": "keith.kalpz@gmailio.com",
      "phone": "(168) 374-6980",
      "status": "active"
    },
    {
      "id": "5",
      "firstName": "Felton",
      "lastName": "Feit",
      "email": "feltonfeit@eggmail.com",
      "phone": "(483) 695-7967",
      "status": "active"
    },
];