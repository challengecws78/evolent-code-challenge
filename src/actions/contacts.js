import * as types from '../utils/types';
import uuid from 'uuid';

export const fetchContacts = contacts => ({
    type: types.FETCH_CONTACTS,
    contacts
});

export const addContact = contact => ({
    type: types.ADD_CONTACT,
    contact: {
        id: uuid(),
        ...contact
    }
});

export const updateContact = (id , contact) => ({
    type: types.UPDATE_CONTACT,
    id,
    contact
});

export const removeContact = (id, contact) => ({
    type: types.REMOVE_CONTACT,
    id,
    contact
});