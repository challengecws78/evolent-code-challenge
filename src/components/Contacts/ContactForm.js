import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ValidatedInput from '../Inputs/ValidatedInput';
import { isEmailValid } from '../../utils/stringHelpers';
export class ContactForm extends Component {
    constructor(props) {
        super(props)
        this.state = {
            firstName: props.firstName || '',
            lastName: props.lastName || '',
            email: props.email || '',
            phone: props.phone || '',
        }
    }
    

    static propTypes = {
        onSubmit: PropTypes.func.isRequired,
        firstName: PropTypes.string,
        lastName: PropTypes.string,
        email: PropTypes.string,
        phone: PropTypes.string
    }
    

    handleChange = ({ name, value }) => {
        this.setState({
            [name]: value
        });
    }

    get getValidation () {
        const { firstName, lastName, email, phone } = this.state;
        const firstNameLength = firstName.length;
        const lastNameLength = lastName.length;
        const emailValid = email.length;
        const phoneLength = phone.length;

        if (firstNameLength > 0 && lastNameLength > 0 && emailValid && phoneLength > 0) {
            return false;
        } else {
            return true;
        }
    }

    onSubmit = e => {
        e.preventDefault();
        const { firstName, lastName, email, phone } = this.state;   
        this.props.onSubmit({
            firstName,
            lastName,
            email,
            phone, 
        });
    }
    
    render() {
        return (
            <div>
                <form className="contact-form" onSubmit={this.onSubmit}>
                    <ValidatedInput
                        label="First Name" 
                        name="firstName"
                        type="text"
                        placeholder="First Name"
                        value={this.state.firstName}
                        onChange={this.handleChange}
                        validate={val => (val ? false : 'First Name is Required')}
                    />
                    <ValidatedInput
                        label="Last Name" 
                        name="lastName"
                        type="text"
                        placeholder="Last Name"
                        value={this.state.lastName}
                        onChange={this.handleChange}
                        validate={val => (val ? false : 'Last Name is Required')}
                    />
                    <ValidatedInput
                        label="Email" 
                        name="email"
                        type="text"
                        placeholder="Email"
                        value={this.state.email}
                        onChange={this.handleChange}
                        validate={val => (isEmailValid(val) ? false : 'Invalid Email')}
                    />
                    <ValidatedInput
                        label="Phone" 
                        name="phone"
                        type="text"
                        placeholder="Phone"
                        value={this.state.phone}
                        onChange={this.handleChange}
                        validate={val => (val ? false : 'Phone is Required')}
                    />
                    <button type="submit" className="btn btn-primary" disabled={this.getValidation}>{this.props.title}</button>
                </form>
            </div>
        );
    }
}



export default ContactForm;