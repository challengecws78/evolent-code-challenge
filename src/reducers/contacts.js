import * as types from '../utils/types';

export default (state=[], action) => {
    switch(action.type) {
        case types.FETCH_CONTACTS:
            return action.contacts
        case types.ADD_CONTACT:
            return [
                ...state,
                action.contact
            ]
        case types.UPDATE_CONTACT:
            return state.map(contact => {
                if(contact.id === action.id){
                    return {
                      ...contact,
                      ...action.contact
                    }
                } else {
                    return contact
                }
            });
        case types.REMOVE_CONTACT:
            return state.map(contact => {
                if(contact.id === action.id){
                    return {
                    ...contact,
                    ...action.contact
                    }
                } else {
                    return contact
                }
            });
        default:
            return state;
    }
}