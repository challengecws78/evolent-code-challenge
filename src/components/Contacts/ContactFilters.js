import React, { Component } from 'react';
import { setContactFilter } from '../../actions/filters';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

export class ContactFilters extends Component {
    handleChange = e => {
        const { name, value } = e.target
        this.props.setContactFilter(name, value);
    }
    render() {
        return (
            <div className="filters">
                <fieldset>
                    <h2>Filters</h2>
                    <div className="sort">
                        <button name="sort" className="btn btn-primary" value="asc" onClick={this.handleChange}>Sort A - Z</button> 
                        <button name="sort" className="btn btn-primary" value="desc" onClick={this.handleChange}>Sort Z - A</button>
                    </div>
                    <label>Search Email</label>
                    <input 
                        name="searchText" 
                        type="text"
                        className="form-control"
                        onChange={this.handleChange}
                        value={this.props.filters.searchText}
                    />
                </fieldset>                
            </div>
        )
    }
}

const mapStateToProps = ({ filters }) => ({ filters });

const mapDispatchToProps = (dispatch) => ({
    setContactFilter: (name, value) => dispatch(setContactFilter(name, value))
});

ContactFilters.propTypes = {
    setContactFilter: PropTypes.func.isRequired,
    filters: PropTypes.object
}

export default connect(mapStateToProps, mapDispatchToProps)(ContactFilters);