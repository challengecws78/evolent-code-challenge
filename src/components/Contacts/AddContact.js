import React, { Component } from 'react';
import { connect } from 'react-redux';
import ContactForm from '../Contacts/ContactForm';
import { addContact } from '../../actions/contacts';
import PropTypes from 'prop-types';

export class AddContact extends Component {
  onSubmit = contact => {
    this.props.onSubmit({
      ...contact,
      status: "active",
    });
    this.props.history.push('/');
  };

  render() {
    return (
      <div className="contact-input">
        <h1>Add Contact</h1>
        <ContactForm
          title="Add Contact"
          onSubmit={this.onSubmit}
        />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  onSubmit: contact => dispatch(addContact(contact))
});

AddContact.propTypes = {
  onSubmit: PropTypes.func.isRequired
}



export default connect(undefined, mapDispatchToProps)(AddContact);