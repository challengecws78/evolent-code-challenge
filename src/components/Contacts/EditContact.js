import React, { Component } from 'react';
import { connect } from 'react-redux';
import ContactForm from './ContactForm';
import { updateContact, removeContact } from '../../actions/contacts';
import PropTypes from 'prop-types';

export class EditContact extends Component {
    onSubmit = (contact) => {
      this.props.updateContact(this.props.contact.id, contact)
      this.props.history.push('/');
    }
    

    onRemove = () => {
        const inactiveContact = {
          ...this.props.contact, 
          status: "inactive"
        };
        this.props.removeContact(this.props.contact.id, inactiveContact);
        this.props.history.push('/');
    }
  
    render() {
      return (
        <div className="contact-input">
            <h1>Edit Contact</h1>
            <ContactForm
                title="Update Contact"
                {...this.props.contact}
                onSubmit={this.onSubmit}
            />
            <button className="btn-danger btn remove" onClick={this.onRemove}>remove</button>
        </div>
      );
    }
  }
  
  const mapStateToProps = ({ contacts }, props) => ({
    contact: contacts.find(contact => contact.id === props.match.params.id)
  });
  
  const matchDispatchToProps = (dispatch) => ({
    updateContact: (id, contact) => dispatch(updateContact(id, contact)),
    removeContact: (id, contact) => dispatch(removeContact(id, contact))
  });
  
  EditContact.propTypes = {
    contact: PropTypes.object.isRequired,
    updateContact: PropTypes.func.isRequired,
    removeContact: PropTypes.func.isRequired
  }
  
  export default connect(mapStateToProps, matchDispatchToProps)(EditContact);