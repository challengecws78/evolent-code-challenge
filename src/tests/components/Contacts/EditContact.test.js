import React from 'react';
import { shallow } from 'enzyme';
import { EditContact } from '../../../components/Contacts/EditContact';
import { contacts } from '../../fixtures/contacts';


let removeContact, updateContact, history, wrapper;

beforeEach(() => {
  updateContact = jest.fn();
  removeContact = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditContact
      updateContact={updateContact}
      removeContact={removeContact}
      history={history}
      contact={contacts[1]}
    />
  );
});

test('should render EditContact', () => {
  expect(wrapper).toMatchSnapshot()
});

test('should handle handle editing a contact', () => {
  wrapper.find('ContactForm').prop('onSubmit')(contacts[1]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(updateContact).toHaveBeenLastCalledWith(contacts[1].id, contacts[1]);
});

test('should handle removing a contact', () => {
  wrapper.find('button').simulate('click');
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(removeContact).toHaveBeenLastCalledWith(contacts[1].id, {...contacts[1], "status": "inactive"});
});
