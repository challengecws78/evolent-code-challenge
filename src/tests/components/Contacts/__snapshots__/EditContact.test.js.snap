// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditContact 1`] = `
<div
  className="contact-input"
>
  <h1>
    Edit Contact
  </h1>
  <ContactForm
    email="chrissymizzell@gmailio.com"
    firstName="Chrissy"
    id="2"
    lastName="Mizzell"
    onSubmit={[Function]}
    phone="5052812342"
    title="Update Contact"
  />
  <button
    className="btn-danger btn remove"
    onClick={[Function]}
  >
    remove
  </button>
</div>
`;
